testing/fil_token_integration/actors/basic_token_actor/src/lib.rs

```diff
@@ -65,6 +66,19 @@ impl FRC46Token<RuntimeError> for BasicToken<'_> {
 
         let ret = hook.call(self.util.msg())?;
 
+        let new_cid = sdk::sself::root().unwrap();
+        let ret = if cid == new_cid {
+            ret
+        } else {
+            // state has changed, update return data with new balances
+            self.util.load_replace(&new_cid)?;
+            TransferReturn {
+                from_balance: self.balance_of(operator)?,
+                to_balance: self.balance_of(to)?,
+                recipient_data: ret.recipient_data,
+            }
+        };
```

Comment:
Seeing this, I'm now wondering about yet another change to the hook return to encourage safety. A token impl should generally re-inspect state to create the return value. There's a possible optimisation where the CID hasn't changed, but the general case is to reload.

So now I think we should _not_ return the ret from the hook, but force the caller to recreate it. But make that easy by adding token methods that create a TransferReturn, MintReturn etc from the current state. The hook return would only contain the recipient data.

So, sketching...

```
let hook_ret = hook.call()?;
self.reload()?; // reloads state only if CID has changed
return self.util.transfer_return(hook_ret);
```

Reply:
I think things get simpler with an approach like that. Might want to rename the underlying functions from `mint` to `mint_init` or something to make clearer that it's a multi-step process? I'll leave that alone for the moment but it's a quick enough change afterwards if we want it in there.

The `reload` in your example probably needs to take the original CID we got when flushing the state before the hook call, but this otherwise looks like a simple enough change. Will try it out once I've addressed other review feedback as it's a more complex change than those.

Reply:
Had a go at it, definitely feels nicer to use when implementing a token.